# Coding Exercises (Appendix 3)

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment. 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need a hint, check the __Hints Section__ at the end of this Notebook. Exercises and Hints are numerated accordingly.

If you need some further help or if you want to check your code, you can also watch the __solutions videos__ or check the __solutions notebook__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

### Exercise 10: Indexing and Slicing DataFrames

7. __Import Pandas__ (as pd), <br>
__Import__ the Cars Dataset (__cars.csv__) and save the DataFrame in the variable __cars__ and<br>
__Inspect/Print__ cars!

In [1]:
import pandas as pd

In [3]:
cars = pd.read_csv('cars.csv')

In [5]:
cars

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0    18.0          8         307.0       130.0    3504          12.0   
1    15.0          8         350.0       165.0    3693          11.5   
2    18.0          8         318.0       150.0    3436          11.0   
3    16.0          8         304.0       150.0    3433          12.0   
4    17.0          8         302.0       140.0    3449          10.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                       name  
0            70     usa  chevrolet chevelle malibu  
1            70     usa          buick skylark 320  
2            70     usa         plymouth satellite  
3            70     usa              amc rebel sst  
4            70     usa                ford torino  
..          ...     ...                        ...  
393          82     usa            ford mustang gl  
394          82  europe                  vw pickup  
395          82     usa              dodge rampage  
396          82     usa                ford ranger  
397          82     usa             chevrolet s-10  

[398 rows x 9 columns]

8. Select the column __"name"__ with the __attribute__ (dot) notation! What is the __name__ of the car with __index label 393__?

In [6]:
cars.name

0      chevrolet chevelle malibu
1              buick skylark 320
2             plymouth satellite
3                  amc rebel sst
4                    ford torino
                 ...            
393              ford mustang gl
394                    vw pickup
395                dodge rampage
396                  ford ranger
397               chevrolet s-10
Name: name, Length: 398, dtype: object

The name is ... ford mustang gl!

9. Apply also the __second alternative__ to select the column __"name"__!

In [7]:
cars["name"]

0      chevrolet chevelle malibu
1              buick skylark 320
2             plymouth satellite
3                  amc rebel sst
4                    ford torino
                 ...            
393              ford mustang gl
394                    vw pickup
395                dodge rampage
396                  ford ranger
397               chevrolet s-10
Name: name, Length: 398, dtype: object

10. __Select__ the car/row at __index position 100__ with the __iloc__ operator (all columns)! How many __cylinders__ does the car have?

In [8]:
cars.iloc[100]

mpg                      18.0
cylinders                   6
displacement            250.0
horsepower               88.0
weight                   3021
acceleration             16.5
model_year                 73
origin                    usa
name            ford maverick
Name: 100, dtype: object

The car has ... 6 cylinders!

11. Select the car/row at __index position 200__ with the __iloc__ operator (and only the column __"name"__)! <br>
What is the __name__ of the car?

In [11]:
cars.iloc[200]["name"]

'ford granada ghia'

It´s a ... ford granada ghia!

12. Select the __last 10 rows__ and the columns __mpg, horsepower, origin__ and __name__ with the __iloc__ operator! __Fill in the Gaps__! <br>
What is the __mpg__ of the __vw pickup__?

In [13]:
cars.iloc[-10:, [0, 3, 7, 8]]

mpg  horsepower  origin                        name
388  26.0        92.0     usa  chrysler lebaron medallion
389  22.0       112.0     usa              ford granada l
390  32.0        96.0   japan            toyota celica gt
391  36.0        84.0     usa           dodge charger 2.2
392  27.0        90.0     usa            chevrolet camaro
393  27.0        86.0     usa             ford mustang gl
394  44.0        52.0  europe                   vw pickup
395  32.0        84.0     usa               dodge rampage
396  28.0        79.0     usa                 ford ranger
397  31.0        82.0     usa              chevrolet s-10

The mpg is ... 44.0!

13. __Import__ the __Cars Dataset__ one more time (save in the variable __cars__) and set the column __"name"__ as the __index__! <br>
__Fill in the Gaps__!

In [ ]:
cars = pd.read_csv(---, --- = "name")

In [ ]:
#run the cell and inspect!
cars

In [ ]:
#run the cell and inspect!
cars.info()

14. __Inspect__ the new Index of the DataFrame by calling the __index__ attribute!

15. Select the __row__ with the index label __"ford ranger"__ (use the __loc__ operator)! What is the __weight__?

The weight is ... 2625!

16. Select the __row__ with the index label __"ford torino"__ and only the columns __"model_year"__ and __"origin"__! <br>
What´s the __model year__?

The model year is ... 70!

17. Select the columns __"mpg"__ and __"weight"__ (all rows) with the __loc__ operator!

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Hints (Spoiler!)

7. pandas as pd, pd.read_csv("filename"), head()

8. cars.---

9. cars["---"]

10. cars.iloc[---]

11. cars.iloc[row index position, column index position]; "name" column is at index position ... 8

12. The last 10 rows = from index position -10 till the end (:); Count the column index positions of the Columns


13. parameter index_col

14. cars.---

15. cars.loc["row label"]

16. cars.loc[---, [---, ---]]

17.  cars.loc[ : , [---, ---]]